File: android/src/main/kotlin/lazebny/io/jc/common/JcCallStateApi.kt
package lazebny.io.jc.common

import android.util.Log
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.EventChannel
import lazebny.io.jc.logic.JcWrapper.JCEvent.JCEvent
import org.greenrobot.eventbus.EventBus
import org.greenrobot.eventbus.Subscribe

interface JcCallStateApi {
    fun microphone(): Boolean
    fun video(): Boolean
    fun speaker(): Boolean
    fun otherMicrophone(): Boolean
    fun otherVideo(): Boolean
    fun callStatus(): String

    companion object {
        private var selfMemberSink: EventChannel.EventSink? = null
        private var otherMemberSink: EventChannel.EventSink? = null
        private var callStatusSink: EventChannel.EventSink? = null
        private var stateApi: JcCallStateApi? = null

        private fun sendSelfMember() {
            val api = stateApi ?: return
            val map = mapOf<String?, Any>(
                "video" to api.video(),
                "microphone" to api.microphone(),
                "speaker" to api.speaker(),
            )
            selfMemberSink?.success(map)
        }

        private fun sendOtherMember() {
            val api = stateApi ?: return
            val map = mapOf<String?, Any>(
                "video" to api.otherVideo(),
                "microphone" to api.otherMicrophone(),
            )
            otherMemberSink?.success(map)
        }

        private fun sendCallStatus() {
            val api = stateApi ?: return
            callStatusSink?.success(api.callStatus())
        }

        @Subscribe
        fun onEvent(event: JCEvent) {
            print("JcCallStateApi onEvent")
            val api = stateApi ?: return
            // do not trigger updates on log events
            val triggerEvents = listOf(
                JCEvent.EventType.CALL_UI,
                JCEvent.EventType.CALL_ADD,
                JCEvent.EventType.CALL_REMOVE,
                JCEvent.EventType.CALL_UPDATE,
                JCEvent.EventType.CAMERA_UPDATE,
            )
            val contains = triggerEvents.contains(event.eventType)
            if (!contains) return

            sendSelfMember()

            sendOtherMember()

            sendCallStatus()
        }

        fun setUp(binaryMessenger: BinaryMessenger, stateApi: JcCallStateApi?) {
            val selfMember =
                EventChannel(binaryMessenger, "lazebny.io.jc/jc_call_state_channel/self")
            val otherMember =
                EventChannel(binaryMessenger, "lazebny.io.jc/jc_call_state_channel/other")
            val callStatus =
                EventChannel(binaryMessenger, "lazebny.io.jc/jc_call_state_channel/status")
            this.stateApi = stateApi
            if (stateApi == null) {
                EventBus.getDefault().unregister(this)
                selfMember.setStreamHandler(null)
                otherMember.setStreamHandler(null)
                return
            }

            EventBus.getDefault().register(this)
            selfMember.setStreamHandler(
                object : EventChannel.StreamHandler {
                    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
                        selfMemberSink = events
                        sendSelfMember()
                    }

                    override fun onCancel(arguments: Any?) {
                        selfMemberSink = null
                    }
                }
            )
            otherMember.setStreamHandler(
                object : EventChannel.StreamHandler {
                    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
                        otherMemberSink = events
                        sendOtherMember()
                    }

                    override fun onCancel(arguments: Any?) {
                        otherMemberSink = null
                    }
                }
            )

            callStatus.setStreamHandler(
                object : EventChannel.StreamHandler {
                    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
                        callStatusSink = events
                        sendCallStatus()
                    }

                    override fun onCancel(arguments: Any?) {
                        callStatusSink = null
                    }
                }
            )

            Log.i("JcSdk", "The channels are set up, update data")
        }

    }
}

File: ios/Classes/common/JcCallStateApi.swift
//
//  JcCallStateApi.swift
//  jc
//
//  Created by Михаил on 30.05.2023.
//

import Foundation
import Flutter

protocol JcCallStateApi {
    func microphone() -> Bool
    
    func video() -> Bool
    
    func speaker() -> Bool
    
    func otherMicrophone() -> Bool
    
    func otherVideo() -> Bool
    
    func callStatus() -> String
}

class JcCallStateApiSetup {
    
    static private var selfMemberSink: FlutterEventSink?
    static private var otherMemberSink: FlutterEventSink?
    static private var callStatusSink: FlutterEventSink?
    static private var stateApi: JcCallStateApi?
    
    private static func sendSelfMember() {
        guard let api = stateApi else { return }
        let map: [String: Any] = [
            "video": api.video(),
            "microphone": api.microphone(),
            "speaker": api.speaker()
        ]
        selfMemberSink?(map)
    }

    private static func sendOtherMember() {
        guard let api = stateApi else { return }
        let map: [String: Any] = [
            "video": api.otherVideo(),
            "microphone": api.otherMicrophone()
        ]
        otherMemberSink?(map)
    }

    private static func sendCallStatus() {
        guard let api = stateApi else { return }
        callStatusSink?(api.callStatus())
    }
    
    @objc static func handleNotification(_ notification: NSNotification) {
        sendSelfMember()
        sendOtherMember()
        sendCallStatus()
    }
    
    static func setUp(binaryMessenger: FlutterBinaryMessenger, stateApi: JcCallStateApi?) {
        let selfMemberChannel = FlutterEventChannel(name: "lazebny.io.jc/jc_call_state_channel/self", binaryMessenger: binaryMessenger)
        let otherMemberChannel = FlutterEventChannel(name: "lazebny.io.jc/jc_call_state_channel/other", binaryMessenger: binaryMessenger)
        let callStatusChannel = FlutterEventChannel(name: "lazebny.io.jc/jc_call_state_channel/status", binaryMessenger: binaryMessenger)
        
        self.stateApi = stateApi
        if stateApi == nil {
            selfMemberChannel.setStreamHandler(nil)
            otherMemberChannel.setStreamHandler(nil)
            callStatusChannel.setStreamHandler(nil)
            NotificationCenter.default.removeObserver(self)
            return
        }
        
        selfMemberChannel.setStreamHandler(StreamHandler(onListen: { sink in
            selfMemberSink = sink
            sendSelfMember()
        }))
        
        otherMemberChannel.setStreamHandler(StreamHandler(onListen: { sink in
            otherMemberSink = sink
            sendOtherMember()
        }))
        
        callStatusChannel.setStreamHandler(StreamHandler(onListen: { sink in
            callStatusSink = sink
            sendCallStatus()
        }))
        
        NotificationCenter.default.addObserver(self, selector: #selector(handleNotification(_:)), name: callNotification, object: nil)
        NotificationCenter.default.addObserver(self, selector: #selector(handleNotification(_:)), name: mediaDeviceNotification, object: nil)
    }
}
